SwiftCompilerSources/Sources/Optimizer/Utilities/AccessUtils.swift

```diff
@@ -0,0 +1,421 @@
+    private var foundBeginAccess: BeginAccessInst? = nil
+    var scope: EnclosingScope? {
+      if let ba = foundBeginAccess {
+        return .scope(ba)
+      } else {
+        guard let accessPath = result else { return nil }
+        return .base(accessPath.base)
+      }
+    }
+
+    private var pointerId = PointerIdentification()
+
+    mutating func start() {
+      result = nil
+      foundBeginAccess = nil
+    }
+
+    struct Path : SimpleWalkingPath {
+      let projectionPath: SmallProjectionPath
+
+      // Tracks whether an `index_addr` instruction was crossed.
+      // It should be (FIXME: check if it's enforced) that operands
+      // of `index_addr` must be `tail_addr` or other `index_addr` results.
+      let indexAddr: Bool
+
+      init(projectionPath: SmallProjectionPath = SmallProjectionPath(), indexAddr: Bool = false) {
+        self.projectionPath = projectionPath
+        self.indexAddr = indexAddr
+      }
+
+      func with(projectionPath: SmallProjectionPath) -> Self {
+        return Self(projectionPath: projectionPath, indexAddr: indexAddr)
+      }
+
+      func with(indexAddr: Bool) -> Self {
+        return Self(projectionPath: projectionPath, indexAddr: indexAddr)
+      }
+
+      func merge(with other: Self) -> Self {
+        return Self(
+          projectionPath: projectionPath.merge(with: other.projectionPath),
+          indexAddr: indexAddr || other.indexAddr
+        )
+      }
+    }
+
+    mutating func rootDef(address: Value, path: Path) -> WalkResult {
+      // Try identifying the address a pointer originates from
+      if let p2ai = address as? PointerToAddressInst,
+         let newAddress = pointerId.getOriginatingAddressOrArgument(p2ai) {
+        return walkUp(address: newAddress, path: path)
+      }
+
+      // If this is a base then we're done
+      if let base = AccessBase(baseAddress: address) {
+        self.result = AccessPath(base: base, projectionPath: path.projectionPath)
+        return .continueWalk
+      }
+
+      // The base is unidentified
+      self.result = nil
+      return .abortWalk
+    }
+
+    mutating func walkUp(address: Value, path: Path) -> WalkResult {
+      if address is IndexAddrInst {
+        // Track that we crossed an `index_addr` during the walk-up
+        return walkUpDefault(address: address, path: path.with(indexAddr: true))
+      } else if path.indexAddr && !canBeOperandOfIndexAddr(address) {
+        // An `index_addr` instruction cannot be derived from an address
+        // projection. Bail out
+        self.result = nil
+        return .abortWalk
+      } else if let ba = address as? BeginAccessInst, foundBeginAccess == nil {
+        foundBeginAccess = ba
+      }
+      return walkUpDefault(address: address, path: path.with(indexAddr: false))
+    }
+  }
+}
+
+struct AccessStoragePathWalker {
+  private var walker = AccessStoragePathWalker()
+  mutating func compute(_ ap: AccessPath) -> [AccessStoragePath] {
```

Comment:
`compute` is a bad name.
Also, please don't use Swift Arrays. Two possibilities how to solve this:
1. Make `walker.origins` a `Stack` and return a custom Sequence here which iterates over the walker's origins.
2. Pass a closure to this function which is called for each found AccessStoragePath.